mypy/semanal.py

```diff
@@ -1288,6 +1291,126 @@ def store_declared_types(self, lvalue: Node, typ: Type) -> None:
+        if isinstance(old_type, TupleType):
+            newtype_class_info = self.build_newtype_typeinfo(name, old_type, old_type.fallback)
+            newtype_class_info.tuple_type = old_type
+        elif isinstance(old_type, Instance):
+            newtype_class_info = self.build_newtype_typeinfo(name, old_type, old_type)
+        else:
+            message = "Argument 2 to NewType(...) must be subclassable (got {})"
+            self.fail(message.format(old_type), s)
+            return
+
+        # If so, add it to the symbol table.
+        node = self.lookup(name, s)
+        # TODO: why does NewType work in local scopes despite always being of kind GDEF?
+        node.kind = GDEF
+        node.node = newtype_class_info
+        call.analyzed = NewTypeExpr(newtype_class_info).set_line(call.line)
+
+    def get_newtype_declaration(self, s: AssignmentStmt) -> Optional[CallExpr]:
+        """Returns the Newtype() call statement if `s` is a newtype declaration
+        or None otherwise."""
+        # TODO: determine if this and get_typevar_declaration should be refactored
+        if len(s.lvalues) != 1 or not isinstance(s.lvalues[0], NameExpr):
+            return None
+        if not isinstance(s.rvalue, CallExpr):
+            return None
+        call = s.rvalue
+        if not isinstance(call.callee, RefExpr):
+            return None
+        callee = call.callee
+        if callee.fullname != 'typing.NewType':
+            return None
+        return call
+
+    def get_newtype_name(self, s: AssignmentStmt) -> Optional[str]:
+        lvalue = cast(NameExpr, s.lvalues[0])
+        name = lvalue.name
+        if not lvalue.is_def:
+            if s.type:
+                self.fail("Cannot declare the type of a NewType declaration", s)
+            else:
+                self.fail("Cannot redefine '%s' as a NewType" % name, s)
+            return None
+        return name
+
+    def check_newtype_args(self, call: CallExpr, name: str, context: Context) -> Optional[Type]:
+        has_failed = False
+        args = call.args
+        if len(args) != 2:
+            self.fail("NewType(...) expects exactly two arguments", context)
+            return None
+
+        # Check first argument
+        if not isinstance(args[0], (StrExpr, BytesExpr, UnicodeExpr)):
+            self.fail("Argument 1 to NewType(...) must be a string literal", context)
+            has_failed = True
+        elif call.arg_kinds[0] != ARG_POS:
+            self.fail("Argument 1 to NewType(...) must be a positional string literal", context)
+            has_failed = True
+        elif cast(StrExpr, call.args[0]).value != name:
+            self.fail("Argument 1 to NewType(...) does not match variable name", context)
+            has_failed = True
+
+        # Check second argument
+        try:
+            value = self.anal_type(expr_to_unanalyzed_type(call.args[1]))
+        except TypeTranslationError:
```

Comment:
Does that come out of anal_type(), expr_to_unanalyzed_type(), or both? If not both I'd like the try/except to be more precise.